Security Documentation & Implementation (#1909)
finos/FDC3: website/docs/api/security.md

---
id: security
sidebar_label: Security & Identity
title: Security & Identity (next)
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

:::info _[@experimental](../fdc3-compliance#experimental-features)_

> 

Reviewer: **Overall document structure**

The ordering of sections means that several concepts are referenced before they are introduced to the reader. Suggested order that builds concepts progressively:

1. Overview / motivation (keep as-is)
2. Desktop Agent Requirements — normative constraint that frames everything else (see separate comment)
3. **Introduce asymmetric cryptography** (public/private keys, JWKS) — before any section that relies on it
4. Trust model (can now reference keys meaningfully)
5. App Identity and Signatures (builds on keys + trust)
6. Encrypted Communications (builds on keys + trust + signatures)
7. User Identity (builds on all of the above)

Please restructure accordingly.

Security and Identity features are experimental additions to FDC3. Limited aspects of their design may change in future.

:::

As FDC3 evolves from desktop containers to web-based implementations, new security challenges arise in open, decentralized environments. This specification defines mechanisms for **application identity verification**, **encrypted communications**, and **user identity sharing** across FDC3-enabled applications.

## Overview

FDC3 Security addresses the following key challenges:

1. **Shift to Web**: Web environments are more hostile than controlled desktop containers, requiring robust identity verification
2. **App Identity**: Applications need verifiable identities to establish trust
3. **User Authentication**: Users need portable identity across heterogeneous applications  
4. **Data Integrity**: Context data requires authenticity guarantees
5. **Scalable Trust**: Moving beyond bilateral trust relationships to circles of trust

The security framework introduces:

- **Digital Signatures** for proving data authenticity and app identity
- **Encrypted Channels** for private communications

> - **JWT-based User Identity** for portable authentication

Reviewer: The Overview section lists challenges and introduced features but the reader has not yet been introduced to the concept of asymmetric encryption (public/private key pairs, JWKS endpoints), which underpins everything that follows. A short conceptual introduction should appear before the challenge list, explaining that the security model is built on asymmetric cryptography — that applications publish their public keys at a stable HTTPS endpoint while keeping private keys in a trusted backend — so that the feature list that follows is comprehensible.

Also, item 5 in the challenge list ("Scalable Trust: Moving beyond bilateral trust relationships to circles of trust") introduces the concept of "circles of trust" without defining it. See comment on the Trust Model section for a discussion of whether this framing is appropriate.

| Context Type | Description |
|-------------|-------------|
| [`fdc3.security.user`](../context/ref/security/User) | User identity with JWT |
| [`fdc3.security.userRequest`](../context/ref/security/UserRequest) | Request for user identity |
| [`fdc3.security.symmetricKeyRequest`](../context/ref/security/SymmetricKeyRequest) | Request for encryption key |
| [`fdc3.security.symmetricKeyResponse`](../context/ref/security/SymmetricKeyResponse) | Encryption key response |
| [`fdc3.security.encryptedContext`](../context/ref/security/EncryptedContextWrapper) | Encrypted context wrapper |

### Intents

| Intent | Input Context | Output Context | Description |
|--------|---------------|----------------|-------------|
| `GetUser` | `fdc3.security.userRequest` | `fdc3.security.user` | Request user identity from an IDP |

## Trust Boundaries

Web applications split into a _front end_ (browser) and _back end_ (server) because servers are trusted and can hold private keys safely. The FDC3 Security API therefore has public and private parts: the front end holds only public keys and delegates signing, verification, and sensitive logic to the back end. 

> The diagram below shows how traffic crosses a WebSocket (or similar secure boundary). The browser uses [`connectRemoteHandlers`](https://github.com/finos/FDC3/blob/main/packages/fdc3-security/src/secure-boundary/ClientSideHandlersImpl.ts); the server runs your [`FDC3Handlers`](https://github.com/finos/FDC3/blob/main/packages/fdc3-security/src/secure-boundary/FDC3Handlers.ts) implementation (for example a subclass of [`DefaultFDC3Handlers`](https://github.com/finos/FDC3/blob/main/packages/fdc3-security/src/secure-boundary/FDC3Handlers.ts)). Each labeled message maps to one interface method.

Reviewer: This section conflates a normative concept (applications should delegate signing and key operations to a trusted backend) with specific reference-implementation details (`FDC3Handlers`, `connectRemoteHandlers`, `setupWebsocketServer` from `@finos/fdc3-security`). Only the concept belongs in the Standard's documentation at this level.

The section title "Trust Boundaries" is also misleading — it describes the frontend/backend split, not trust between applications. A title like "Trusted Backend Contract" better reflects the content.

The name `FDC3Handlers` is non-specific — it reads like a generic message-handling interface rather than something specific to the trusted-backend security concern. A more descriptive name such as `SecurityBackendHandlers` or `TrustedBackendHandlers` would better convey its role.

The opening prose is also confusing: it describes `FDC3Handlers` as the contract the **trusted backend** implements, but the JSDoc in the code block then says:

> **`ClientSideHandlersImpl`** — Implements `FDC3Handlers` on the **low-trust** side as a remote stub.

This implies `FDC3Handlers` is implemented on the *client* side (as a stub), not the server. The relationship between the two sides needs to be stated clearly.

Suggested replacement for lines 53–57:

```suggestion
## Trusted Backend Contract

Web applications are split into a *front end* (browser context) and a *back end* (server). Because servers can hold private keys securely, all cryptographic operations — signing, key generation, decryption — MUST be performed in the trusted backend. The frontend holds only public keys and delegates sensitive operations to the backend over a secure channel (e.g. a WebSocket).

The contract between the frontend and backend is application-defined, but must cover at minimum:
- Signing outbound context objects
- Generating or holding symmetric encryption keys
- Unwrapping JWE-wrapped keys received from other applications
```

The `FDC3Handlers` detail (lines 84 onwards) should be moved into, or referenced from, the existing `:::tip` callout.

   * strings so both sides agree on the payload shape.   
   *
   * **Samples**
   * - `signing-intent-example.ts` — shows how you can implement `sign-context` when raising an intent so the browser
   *   never sees the private key.
   * - `frontend-encrypted-channel-example.ts` — shows how you can implement `sign-context` and
   *   `unwrap-symmetric-key` on the receiver backend so key-request signing and symmetric-key
   *   unwrapping stay off the frontend (`fdc3.security.symmetricKeyResponse` in, unwrapped JWK out).
   *
   * @param purpose — Tells the server which operation this call represents; defined by your app.
   * @param o — The payload for that operation (for example `{ context }` or an object compatible with
   *   `fdc3.security.symmetricKeyResponse`).
   */
  exchangeData(purpose: string, o: object): Promise<object | void>;
}
```

#### Example 1: Bridging a channel with `handleRemoteChannel`

The front end keeps the live `Channel` from the Desktop Agent; `handleRemoteChannel` mirrors it to the back end so signing can run where private keys live. Broadcasts initiated on the server flow through the proxy to the real channel.

> 

Reviewer: Both sequence diagrams in this section have a participant ordering issue. The App Frontend mediates between the App Backend and the Desktop Agent — all Desktop Agent communication passes through the frontend. The current ordering places `DA` between `FEA` and `BEA` in Example 2, which implies the Desktop Agent communicates directly with the backend, which is not the case.

> 
> ```mermaid
> sequenceDiagram
>     participant FEA as App Front End
>     participant DA as Desktop Agent
>     participant BEA as App Back End
> 
>     DA-->>FEA: Channel
>     FEA->>BEA: handleRemoteChannel(purpose, channel) over WebSocket
>     BEA->>BEA: broadcast signed context
>     Note over BEA: sign context on the back end
>     BEA-->FEA: proxied channel ops
>     FEA->>DA: channel broadcast (with signed context)
>     Note over DA: context delivered to other apps / listeners
> ```

Reviewer: Reorder diagram and adjust notes - Watch out when applying this suggestion as Github tends to clip off the final triple backticks

```suggestion
```mermaid
sequenceDiagram
    participant BEA as App Back End
    participant FEA as App Front End
    participant DA as Desktop Agent

    DA-->>FEA: Channel
    FEA->>BEA: handleRemoteChannel(purpose, channel) over WebSocket
    
    Note over BEA: back end decides to broadcast
    BEA->>BEA: sign context on the back end
    BEA-->>FEA: Signed context passed to front end
    FEA->>DA: channel broadcast (with signed context)
    Note over DA: context delivered to other apps / listeners
```
```

#### Example 2: Resolving intents with `remoteIntentHandler`

> App A registers an intent listener in the browser using a handler obtained from its back end. When App B raises the intent, the Desktop Agent delivers it to App A’s front end, which forwards handling to **App Back End A** and returns the result.
> 
> ```mermaid
> sequenceDiagram
>     participant FEA as App Front End A
>     participant BEA as App Back End A
>     participant DA as Desktop Agent
>     participant FEB as App Front End B
>     Note over FEA:  App A wants to handle MyIntent
>     FEA->>BEA: calls remoteIntentHandler 'MyIntent'
>     BEA-->>FEA: returns remoteIntentHandler
>     FEA->>DA: addIntentListener('MyIntent', remoteIntentHandler)
>     FEB->>DA: raiseIntent('MyIntent', context)
>     DA->>BEA: deliver intent to listener
>     BEA->>DA: handle intent, perhaps return signed or encrypted context
>     DA-->>FEB: resolution
> ```

Reviewer: Watch out when applying this suggestion as Github tends to clip off the final triple backticks

```suggestion
```mermaid
sequenceDiagram
    participant BEA as App Back End A
    participant FEA as App Front End A
    participant DA as Desktop Agent
    participant FEB as App Front End B
    Note over FEA: App A wants to handle MyIntent
    FEA->>BEA: calls remoteIntentHandler 'MyIntent'
    BEA-->>FEA: returns remoteIntentHandler
    FEA->>DA: addIntentListener('MyIntent', remoteIntentHandler)
    FEB->>DA: raiseIntent('MyIntent', context)
    DA->>FEA: deliver intent to listener
    FEA->>BEA: forward to backend handler
    BEA-->>FEA: signed or encrypted result
    FEA->>DA: return result
    DA-->>FEB: resolution
```
```

</TabItem>
</Tabs>

> Working samples (signed broadcasts, encrypted channels, identity, intents) live under [`packages/fdc3-security/samples/`](https://github.com/finos/FDC3/tree/main/packages/fdc3-security/samples).

Reviewer: References to the reference implementation should be wrapped in a `:::tip` callout to make clear they are not normative. This line is currently plain prose:

> Working samples (signed broadcasts, encrypted channels, identity, intents) live under [`packages/fdc3-security/samples/`](https://github.com/finos/FDC3/tree/main/packages/fdc3-security/samples).

Suggested replacement:

```suggestion
:::tip Reference implementation samples
Working samples (signed broadcasts, encrypted channels, identity, intents) live under [`packages/fdc3-security/samples/`](https://github.com/finos/FDC3/tree/main/packages/fdc3-security/samples).
:::
```

### Public / Private Keys

To enable verification, applications publish their public keys at a stable HTTPS endpoint. The conventional location is `/.well-known/jwks.json` on the application's origin—a path established by [RFC 8615](https://datatracker.ietf.org/doc/html/rfc8615) for well-known URIs. The URL serves both as the key delivery mechanism and as an identifier for the publisher: the `jku` in a JWS header points to this endpoint so receivers know where to fetch the verification key.

:::note
Keys in the JWKS MUST include a `kid` (key ID) so signers can reference them in JWS headers and receivers can select the correct key for verification.
:::

1. Generate a public/private key pair

> 2. Publish the public key at an HTTPS endpoint as a [JSON Web Key Set (JWKS)](https://datatracker.ietf.org/doc/html/rfc7517).  The URL on which the JWKS identifies the entity of the publisher. 

Reviewer: This section appears before asymmetric encryption has been introduced as a concept. It needs:

1. An introductory sentence explaining *why* applications need a public/private key pair in this context — i.e. so that other applications can verify their signatures and encrypt data that only they can read.
2. A lead-in sentence before the numbered list, e.g. "To participate in the FDC3 security model, an application should:"

Without these the numbered steps at lines 228–229 read as instructions dropped in with no context.

### Key Management

- Private keys MUST be stored securely and never transmitted
- JWKS endpoints MUST be served over HTTPS with valid certificates
- Key rotation SHOULD be performed periodically
- Old keys SHOULD remain available for verification during transition periods
- Applications SHOULD log authentication and authorization decisions for audit

> - Handle unsigned contexts gracefully with appropriate user prompts

Reviewer: The "Key Management" section lacks an introductory sentence. The bullet list of requirements should be preceded by a sentence explaining the scope and purpose of this section, for example: "Applications participating in the FDC3 security model are responsible for the following key management practices:". Without that framing the bullets read as a disconnected checklist.

## Trust Model

> Trust is determined and enforced by applications, not by the Desktop Agent. Each application defines its own circle of trust. Desktop Agents are untrusted intermediaries: they route context and metadata but MUST NOT be relied upon for cryptographic operations or trust decisions. Application front-ends (browser contexts) are untrusted: private keys MUST NOT be sent to or stored in front-end code; signing and other sensitive operations MUST be performed in a trusted back-end.

Reviewer: "Trust" is used as a security term throughout the document but is never defined. This section should open with a brief definition: in this context, *trust* means that a receiving application has decided that a particular sender's identity claims and signed data are credible and should be acted upon.

The phrase "circle of trust" (also in line 26) is worth reconsidering. What is being established here is not a circle — which implies bidirectional mutual membership — but rather a unilateral *allowlist*: each application independently decides which other applications it trusts. The concept is fully captured by "allowlist" or "trusted senders list" and the additional term "circle of trust" may mislead readers into thinking there is a shared membership or negotiation mechanism involved.

  'https://app-a.example.com/.well-known/jwks.json',
  'https://app-b.example.com/.well-known/jwks.json',
  'https://data-provider.example.com/.well-known/jwks.json',
]);

const allowListFunction = (jku: string, iss?: string): boolean => {
  // jku: JWKS URL from the signer's JWS/JWT header (where to fetch public keys)
  // iss: issuer claim from JWT payload (only passed when verifying JWT tokens, e.g. for user identity)
  return TRUSTED_JWKS_URLS.has(jku);
};
```

```typescript
// Example: using iss for JWT verification (e.g. user identity)
// Restrict which issuers are trusted per JWKS URL—useful when an IdP hosts multiple issuers
const TRUSTED_ISSUERS_BY_JKU: Record<string, Set<string>> = {
  'https://idp.example.com/.well-known/jwks.json': new Set(['https://idp.example.com', 'tenant-a.idp.example.com']),
  'https://enterprise-sso.corp.com/.well-known/jwks.json': new Set(['https://enterprise-sso.corp.com']),
};

const allowListFunction = (jku: string, iss?: string): boolean => {
  if (!TRUSTED_ISSUERS_BY_JKU[jku]) return false;
  // For detached JWS (context signing), iss is undefined—trust jku alone
  if (!iss) return true;
  return TRUSTED_ISSUERS_BY_JKU[jku].has(iss);
};

> ```

Reviewer: Two issues with the `allowListFunction` examples:

1. Neither function has a JSDoc comment explaining its parameters and return value.
2. They are presented as plain prose but relate directly to the `@finos/fdc3-security` reference implementation API. They should be wrapped in a `:::tip` callout or clearly labelled as "Example implementation" to maintain the separation between the standard's concepts and the library's API. As written they appear normative.

Suggested JSDoc to add to both examples:

```typescript
/**
 * Determines whether a signing application is trusted by this application.
 * @param jku - The JWKS URL from the signer's JWS protected header, identifying the signer.
 * @param iss - The issuer claim from a JWT payload; only present when verifying user identity tokens.
 * @returns true if the signer should be trusted; false otherwise.
 */
```

## App Identity and Signatures

### Signing Context Data

Applications can sign the context objects they broadcast using their private key. This allows receiving applications to verify:

1. **Origin**: Which application sent the data
2. **Integrity**: Whether the data was tampered with in transit

### Signature Metadata

> When a context is signed, the signature is provided in metadata (via [`AppProvidableContextMetadata`](ref/Metadata#contextmetadata) on broadcast/raiseIntent, or [`ContextMetadata`](ref/Metadata#contextmetadata) when received), not on the context object itself.

Reviewer: The following sentence conflates two distinct types with different roles:

`AppProvidableContextMetadata` is the type of the optional `metadata` *parameter* that a sending application passes to `broadcast()` or `raiseIntent()`. `ContextMetadata` is the type of the metadata *argument* delivered to a `ContextHandler` or `IntentHandler` — it extends `AppProvidableContextMetadata` with additional Desktop Agent-provided fields such as `source`.

Suggested replacement:

```suggestion
When signing a context, the sender includes the `signature` and `antiReplay` fields in the `metadata` parameter passed to `broadcast()` or `raiseIntent()` (typed as [`AppProvidableContextMetadata`](ref/Metadata#appprovidablecontextmetadata)). The signature is therefore carried separately from the context object itself. When a receiving application's [`ContextHandler`](ref/Types#contexthandler) or [`IntentHandler`](ref/Types#intenthandler) is invoked, those fields are present in the [`ContextMetadata`](ref/Metadata#contextmetadata) argument alongside the Desktop Agent-provided `source` field.
```

| Context Metadata field | Description |
|------------------------|-------------|
| `signature` | The detached JWS (`protected` + `signature`) |

> | `antiReplay` | Claims (`iat`, `exp`, `jti`) used for replay detection; must be included when signing |

Reviewer: The `antiReplay` field listed here does not appear in `website/docs/api/ref/Metadata.md`. The reference documentation for `AppProvidableContextMetadata` and `ContextMetadata` needs to be updated to include `antiReplay` alongside `signature`.

### Flow Diagram

```mermaid
sequenceDiagram
    participant SBE as Sender (trusted back end)
    participant SFE as Sender (front end)
    participant DA as Desktop Agent
    participant R as Receiver

    Note over SBE: Build antiReplay (iat, exp, jti)
    Note over SBE: Canonicalize { context, antiReplay }
    Note over SBE: Sign with private key → detached JWS in metadata
    Note over SBE: Serve JWKS at jku (HTTPS)

    SBE->>SFE: Relay signed message (context + signature + antiReplay)
    SFE->>DA: broadcast / raiseIntent (context + metadata)
    Note over DA: Untrusted relay — forward unchanged
    DA->>R: context + metadata

    Note over R: Decode protected header (alg, jku, kid, …)
    R->>SBE: HTTPS: fetch JWKS (jku)
    SBE-->>R: JWKS / verification key

    Note over R: Reconstitute compact JWS, verify signature
    Note over R: Validate iat, antiReplay.exp, jti
    Note over R: allowListFunction(jku) → authenticity.trusted

    Note over R: Expose context + ContextMetadata.authenticity to app

> ```

Reviewer: The flow diagram references `antiReplay`, `jku`, `kid`, `JWS`, and `allowListFunction` before any of these are defined in the sections that follow (Signature Structure, Checking the Signature, etc.). 

The "App Identity and Signatures" section would be easier to follow if reordered as:
1. Signing Context Data (keep as-is)
2. Signature Metadata (field table — lines 288–295)
3. Signature Structure (lines 350–361)
4. Generating The Signature (lines 342–348)
5. Flow Diagram (can now reference already-defined concepts)
6. Example (lines 327–341)
7. Checking the Signature (lines 363–371)
8. Authenticity Metadata (lines 373–416)

| Header field | Description |
|--------------|-------------|
| `alg` | Signature algorithm (e.g., `EdDSA`) |
| `jku` | URL of the JWKS containing the public key for verification |
| `kid` | Key identifier within the JWKS |
| `iat` | Issued-at time (Unix timestamp), prevents replay |
| `exp` | Expiration time (Unix timestamp) |
| `jti` | Unique token ID for replay protection |

### Checking the Signature

To verify a signature, the receiver:

1. Extracts `alg`, `jku`, `kid`, and `iat` from the JWS protected header.
2. Resolves the public key from the `jku` JWKS URL (via a resolver or [JOSE](https://github.com/panva/jose) remote JWKS).
3. Reconstitutes the full JWS: canonicalizes `{ context, antiReplay }`, base64url-encodes it as the payload, and forms `header.payload.signature`. Uses `compactVerify` (or equivalent) to verify the signature.
4. Validates freshness (`iat`), context expiry (`antiReplay.exp`), and anti-replay claims (`jti`).
5. Populates the `authenticity` object in context metadata.

### Authenticity Metadata

> When a signed context is received, the FDC3 security layer verifies the signature and populates the `authenticity` field in [`ContextMetadata`](ref/Metadata#contextmetadata):

Reviewer: The statement:

> When a signed context is received, the FDC3 security layer verifies the signature and populates the `authenticity` field in `ContextMetadata`.

is not quite right. `ContextMetadata` is the type delivered by the Desktop Agent to a `ContextHandler` or `IntentHandler` — it should not be modified after the fact by a library layer. The correct model appears to be:

1. The application passes the received `ContextMetadata` into a validation function provided by the security library.
2. That function returns a result object (corresponding to the `SecurityMetadata` interface defined in `ContextMetadata.ts`) containing `authenticity`, `antiReplay`, and `encryption` fields.

Additionally, `api.schema.json` currently includes an `authenticity` field directly on `ContextMetadata`. If `authenticity` is a library-computed result rather than a wire field, it should not appear in the wire schema.

The term "FDC3 security layer" used here and throughout the document also blurs the line between the standard and the reference implementation. Suggested alternatives:
- For normative flows: "the receiving application's security implementation"  
- For `@finos/fdc3-security`-specific behaviour: wrap in a `:::tip` callout and name the library explicitly

<Tabs groupId="lang">
<TabItem value="ts" label="TypeScript/JavaScript">

```ts
fdc3.addContextListener("fdc3.instrument", (context, metadata) => {
  const auth = metadata?.authenticity;

  if (!auth?.signed) {
    // No signature present - treat as untrusted
    console.warn("Received unsigned context");
    handleUntrustedContext(context);
  } else if (!auth.valid) {
    // Signature present but verification failed (tampered, stale, or bad key)
    console.warn("Signature verification failed", auth.errors);
    rejectContext(context);
  } else if (!auth.trusted) {
    console.warn(`Untrusted context from: ${auth.jku}`);
    promptUserForConfirmation(context, auth.jku);
  } else {   
    // this is trusted - continue without user intervention
    processVerifiedInstrument(context);
  }
});
```

</TabItem>
</Tabs>

## Encrypted Communications

### Private Channel Encryption

Applications communicating over a [`PrivateChannel`](ref/PrivateChannel) can negotiate encryption to ensure their communications remain confidential. This is particularly important when sharing sensitive data such as positions, pricing, or user information.

### Symmetric Key Exchange

> Encryption uses a symmetric key (e.g. AES-GCM) created by the channel owner and distributed via [JWE](https://datatracker.ietf.org/doc/html/rfc7516). See [encrypted-private-channel-example.ts](pathname:///packages/fdc3-security/samples/encrypted-private-channel-example.ts) for a working flow.

Reviewer: The "Encrypted Communications" and "Symmetric Key Exchange" sections use JWE (JSON Web Encryption) without defining it, and do not explain why symmetric key exchange is preferred over encrypting each message directly with the recipient's public key.

The section needs an introductory passage before line 420, for example:

> While digital signatures (described above) prove *who* sent a message, they do not prevent the Desktop Agent or other intermediaries from reading it. To keep context data confidential, applications can encrypt it so that only the intended recipient can decrypt it.
>
> **Why symmetric key exchange?** Encrypting every message directly with the recipient's asymmetric public key (e.g. RSA or ECDH) is computationally expensive and impractical for a stream of context messages. The approach used here mirrors TLS: asymmetric cryptography is used once to securely exchange a *symmetric key*, then the much cheaper symmetric cipher (e.g. AES-GCM) encrypts the data stream. A [JSON Web Encryption (JWE)](https://datatracker.ietf.org/doc/html/rfc7516) token wraps (encrypts) the symmetric key using the recipient's public key, ensuring only the recipient can unwrap it.

    Rx->>Sx: Raises `ViewInstrument` Intent
    note right of Sx: Generate random symmetric key K
    note right of Sx: Create private channel C
    Sx->>Rx: Intent result: private channel C
    note left of Rx: Join channel C
    Rx->>RxCtx: addContextListener for `fdc3.instrument`
    Rx->>RxKey: addContextListener for `symmetricKeyResponse`
    Sx->>SxReq: addContextListener for `symmetricKeyRequest`
    note right of Sx: New instrument snapshot
    note right of Sx: Encrypt fdc3.instrument with K, sign (JWS)
    Sx->>RxCtx: Broadcast fdc3.security.encryptedContext (originalType fdc3.instrument)
    note right of RxCtx: Payload is JWE — need K
    note right of RxCtx: Verify context JWS (sender)
    RxCtx->>SxReq: Broadcast fdc3.security.symmetricKeyRequest (signed)
    note right of SxReq: Verify request JWS, fetch Receiver JWKS from jku
    note right of SxReq: Wrap K for Receiver (JWE), sign response (JWS)
    SxReq->>RxKey: Broadcast fdc3.security.symmetricKeyResponse
    note right of RxKey: Verify response JWS, unwrap K (private key)
    note right of RxKey: K now held for channel C
    RxKey->>RxCtx: decrypt instrument
    RxCtx->>RxCtx: process context handler

> ```

Reviewer: Two issues in this flow diagram:

1. **Participant names**: "Receiver App" and "Broadcaster App" are functional descriptions rather than the role-descriptive or neutral names used elsewhere in FDC3 examples. Given the use case (a buy-side app requesting a price stream from a sell-side platform), names like "Buy-side App" / "Market Data App", or simply "App A" / "App B", would be clearer.

2. **Intent name**: `ViewInstrument` (line 443) is used as the example intent. Per the naming conventions in `website/docs/intents/spec.md`, the `View` prefix is for intents that display content to the user — not for requesting a data stream. A `Get`-style intent would be more appropriate. There is no existing intent that fits precisely; see the top-level review comment for a suggestion to introduce `GetPricingStream` / `GetPrice`.

#### Context Types

| Type | Description |
|------|-------------|
| [`fdc3.security.symmetricKeyRequest`](../context/ref/security/SymmetricKeyRequest) | Request for the channel symmetric key (optional `id.kid`). Must be signed. |
| [`fdc3.security.symmetricKeyResponse`](../context/ref/security/SymmetricKeyResponse) | Response containing `wrappedKey` (JWE) and `id.{kid,pki}`. Must be signed. |

> | [`fdc3.security.encryptedContext`](../context/ref/security/EncryptedContextWrapper) | Wrapper with `encryptedPayload` (JWE); `originalType` and `id.kid` preserved for routing. |

Reviewer: The process of wrapping a context object in `fdc3.security.encryptedContext` is not defined here or in the `EncryptedContextWrapper` reference doc. Implementors need to know:

- Which fields of the original context are preserved in plaintext in the wrapper (`type` → `originalType`, `id.kid`)
- Which fields are included in the encrypted payload (all remaining fields? The full original context?)
- The exact format of `encryptedPayload` (JWE compact serialization?)
- How a recipient identifies and routes an encrypted context before decryption (by checking `type === 'fdc3.security.encryptedContext'`)

Without this, independent implementations cannot interoperate correctly.

:::tip

#### Javascript Examples

See: 
- https://github.com/finos/FDC3/blob/main/packages/fdc3-security/samples/backend-encrypted-channel-example.ts
- https://github.com/finos/FDC3/blob/main/packages/fdc3-security/samples/frontend-encrypted-channel-example.ts

::: 

## User Identity

In a multi-app FDC3 environment, applications often need to know who the user is—for personalization, access control, audit trails, or to share session state across tools. Instead of each app authenticating the user separately, an **Identity Provider (IDP)** can issue a portable, verifiable identity that any trusted app can consume. The IDP signs a JWT containing user claims; receiving apps verify the signature and trust the identity. 

> **Audience scoping** ensures each JWT is bound to a specific requesting application (`aud`), so if a token is intercepted, it cannot be reused by another app. This enables single sign-on (SSO) across FDC3 applications while keeping tokens narrowly scoped.

Reviewer: Two issues in this introductory paragraph:

1. **IDP terminology**: The term "Identity Provider (IDP)" is likely to mislead readers familiar with OAuth 2.0 / OIDC, where an IdP is a centralised service with redirect flows and token endpoints. In this context the "IDP" is simply another FDC3 application that can vouch for the current user's identity. Suggested alternative on first use: "identity provider app — an FDC3 application that can authenticate the current user and issue signed identity tokens (not to be confused with an OAuth 2.0 / OIDC identity provider; no redirect flows or token endpoints are involved)". All subsequent uses of "IDP" should use "identity provider app" consistently.

2. **`aud` not introduced**: The claim `aud` is referenced in line 488 before being defined. `aud` is a standard JWT claim ([RFC 7519 §4.1.3](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.3)) meaning "audience" — the intended recipient of the token. The paragraph should briefly explain this and clarify what value an application should use (the `UserRequest` example uses a URL, but the document does not explain why a URL rather than, say, an `appId` from the App Directory).

### The User Context Type

The [`fdc3.security.user`](../context/ref/security/User) context type represents a verified user identity:

```typescript
{
  type: "fdc3.security.user",
  wrappedJwt: "--example-jwt-token--but-wrapped-in-the-public-key-of-the-requester--"
}

> ```

Reviewer: The `wrappedJwt` field is not explained before the example. The concept needs a brief introduction: the JWT is *encrypted* (wrapped as a JWE) using the requesting application's public key, so that only the requesting application — which holds the corresponding private key — can decrypt and read it. This prevents the token from being read by the Desktop Agent or any other application that may observe the intent result.

The dummy value `"--example-jwt-token--but-wrapped-in-the-public-key-of-the-requester--"` is currently carrying this explanation implicitly, which is not sufficient — the prose before the code block should explain it explicitly.

### JWT Token Structure

The `wrappedJwt` field contains a signed JSON Web Token.  Once you decrypt this with your application's private key, it will have the following structure:

**Header:**
```json
{
  "alg": "EdDSA",
  "jku": "https://idp.example.com/.well-known/jwks.json",
  "kid": "key-1"
}
```

**Payload:**
```json
{
  "iss": "https://idp.example.com",
  "sub": "john.doe@example.com",
  "aud": "https://requesting-app.example.com",
  "exp": 1739750400,
  "iat": 1739746800,
  "jti": "unique-token-id-123"
}
```

| Claim | Description |
|-------|-------------|
| `iss` | Issuer - the identity provider that created the token |
| `sub` | Subject - the user's unique identifier |
| `aud` | Audience - the specific application this token was issued for |
| `exp` | Expiration time (Unix timestamp) |

> | `iat` | Issued at time (Unix timestamp) |

Reviewer: The `exp` and `iat` claims use Unix timestamps (seconds since epoch), whereas `ContextMetadata.timestamp` elsewhere in FDC3 uses ISO 8601-1:2019 format. This inconsistency is intentional — Unix timestamps (NumericDate) are mandated for these claims by [RFC 7519 §2](https://datatracker.ietf.org/doc/html/rfc7519#section-2) — but it should be noted explicitly so readers understand it is standards-driven rather than an oversight, e.g.:

> Note: `exp` and `iat` use Unix epoch seconds (NumericDate as defined in [RFC 7519](https://datatracker.ietf.org/doc/html/rfc7519)), rather than the ISO 8601 format used elsewhere in FDC3.

| `jti` | JWT ID - unique identifier for this token (replay prevention) |

The token is scoped to a specific application (`aud`) to prevent token reuse if leaked.

### Requesting User Identity

1. The requester raises the [`GetUser`](../intents/ref/GetUser.md) intent with a signed [`fdc3.security.userRequest`](../context/ref/security/UserRequest) context. The context includes an `aud` field (the audience claim for the returned JWT—typically the requesting app's URL). The request must be signed to prevent forgery by third parties.

2. The IDP verifies the requester's signature on the request.

3. The IDP creates a JWT containing the claims described above, scoped to the requested audience.

4. The IDP wraps the JWT with the requester's public key and returns the `fdc3.security.user` context as the intent result.

> 5. The requester decrypts the wrapped JWT with their private key and verifies the JWT signature using the IDP's public keys (from `{idpBaseUrl}/.well-known/jwks.json`). It also verifies expiry and must verify that the `jti` has not been used recently (i.e. in the expiry window)

Reviewer: This section consistently refers to "The IDP" as if it is a distinct system component rather than another FDC3 application. All uses of "IDP" here should be updated to "identity provider app" (consistent with the suggestion in the User Identity intro comment). The flow diagram at line 552 uses "IDP" as a participant label and should similarly be updated, e.g. to "Identity Provider App".

```mermaid
sequenceDiagram
    participant Application
    participant IDP
    Application->>IDP: GetUser Intent Request
    note right of IDP: Verify requesting application signature
    note right of IDP: Generate scoped JWT
    note right of IDP: Wrap JWT with requesting app's public key
    IDP->>Application: GetUser Intent Reply
    note left of Application: Unwrap JWT
    note left of Application: Verify JWT
    note left of Application: Authenticate user
```

### Token Security

- JWT tokens are scoped to specific audiences to prevent misuse if leaked
- Short expiration times reduce the window for token theft attacks
- Unique token identifiers (jti) must be used to prevent token theft attacks
- Tokens SHOULD be transmitted over encrypted channels when possible

## Desktop Agent Requirements

> No changes are expected of Desktop Agents to work with security - just that they pass on context and metadata untampered-with.

Reviewer: This section contains one of the most important normative statements in the document but is buried at the very end and uses informal language. It should be:

1. Moved to immediately after the Overview section, before "Trust Boundaries" — it is a fundamental constraint on Desktop Agent implementors that frames the entire security model.
2. Rewritten using RFC 2119 MUST language consistent with the rest of the standard, for example:

> Desktop Agents MUST forward context objects and their associated metadata to receiving applications without modification. Desktop Agents MUST NOT inspect, strip, or alter `signature`, `antiReplay`, or any other security-related metadata fields. The security model defined in this document relies on this guarantee: cryptographic verification is performed end-to-end between applications, not by the Desktop Agent.

---
id: security
sidebar_label: Security & Identity
title: Security & Identity (next)
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

:::info _[@experimental](../fdc3-compliance#experimental-features)_

Security and Identity features are experimental additions to FDC3. Limited aspects of their design may change in future.

:::

As FDC3 evolves from desktop containers to web-based implementations, new security challenges arise in open, decentralized environments. This specification defines mechanisms for **application identity verification**, **encrypted communications**, and **user identity sharing** across FDC3-enabled applications.

> The security model is built on **asymmetric cryptography**: each participating application holds a private key (kept secret) and publishes a corresponding public key at a stable HTTPS endpoint. Other applications use that public key to verify the application's digital signatures and to encrypt data that only it can read. All cryptographic operations that involve a private key MUST be performed in a trusted backend (server), not in the browser frontend.

Reviewer: This introduction can be improved - it shouldn't refer so.heavily to web as thats not really what gives rise to the need. Rather its the whole desktop agent concept

```suggestion
In the FDC3 model, applications communicate through a Desktop Agent rather than directly with each other. While Desktop Agents are responsible for verifying the identity of the applications they connect (for example, by launching applications from known URLs and validating message origins against directory records), applications themselves have no direct relationship with their counterparts. Because the FDC3 standard aims to build an ecosystem of interoperable applications that can run under any conformant Desktop Agent, applications cannot necessarily know or trust the specific Desktop Agent they are running under and must treat it as an untrusted intermediary when it comes to verifying the identity of other applications or protecting sensitive data in transit.

This gives rise to the need for applications to independently establish the identity of the applications they are communicating with and to secure their communications over a potentially untrusted channel. The security model is built on **asymmetric cryptography**: each participating application holds a private key (kept secret) and publishes a corresponding public key at a stable HTTPS endpoint. Other applications use that public key to verify the application's digital signatures and to encrypt data that only it can read.

This specification defines mechanisms for **application identity verification**, **encrypted communications**, and **user identity sharing** across FDC3-enabled applications.
```

## Overview

FDC3 Security addresses the following key challenges:

1. **Shift to Web**: Web environments are more hostile than controlled desktop containers, requiring robust identity verification
2. **App Identity**: Applications need verifiable identities to establish trust
3. **User Authentication**: Users need portable identity across heterogeneous applications  
4. **Data Integrity**: Context data requires authenticity guarantees

> 5. **Scalable Trust**: Moving beyond bilateral trust relationships to independently maintained allowlists

Reviewer: Key challenges can be improved:

```suggestion
1. **Untrusted Channel**: Applications communicate through a Desktop Agent they may not control or fully trust, requiring end-to-end identity verification
2. **App Identity**: Applications need verifiable identities to establish trust with their counterparts independently of the Desktop Agent
3. **User Authentication**: Users need portable identity across heterogeneous applications  
4. **Data Integrity**: Context data requires authenticity guarantees
5. **Data Confidentiality**: Sensitive data must be protected from the Desktop Agent and other observers
6. **Scalable Trust**: Moving beyond bilateral trust relationships to independently maintained allowlists
```

The security framework introduces:

- **Digital Signatures** for proving data authenticity and app identity
- **Encrypted Channels** for private communications
- **JWT-based User Identity** for portable authentication

### Context Types

The following context types support security features:

| Context Type | Description |
|-------------|-------------|
| [`fdc3.security.user`](../context/ref/security/User) | User identity with JWT |
| [`fdc3.security.userRequest`](../context/ref/security/UserRequest) | Request for user identity |
| [`fdc3.security.symmetricKeyRequest`](../context/ref/security/SymmetricKeyRequest) | Request for encryption key |
| [`fdc3.security.symmetricKeyResponse`](../context/ref/security/SymmetricKeyResponse) | Encryption key response |

> | [`fdc3.security.encryptedContext`](../context/ref/security/EncryptedContextWrapper) | Encrypted context wrapper |
> 
> ### Intents
> 
> | Intent | Input Context | Output Context | Description |

Reviewer: ```suggestion
### Intents

The following intents support user identity features, other Get* intents may be used to create requests for encrypted channels:

| Intent | Input Context | Output Context | Description |
```

|--------|---------------|----------------|-------------|

> | `GetUser` | `fdc3.security.userRequest` | `fdc3.security.user` | Request user identity from an identity provider app |

Reviewer: ```suggestion
| [`GetUser`](../intents/ref/GetUser) | [`fdc3.security.userRequest`](../context/ref/security/UserRequest) | [`fdc3.security.user`](../context/ref/security/User) | Request user identity from an identity provider app |
```
